R/req-promise.R

```diff
@@ -0,0 +1,154 @@
+#' [req_perform_parallel()] function, it:
+#'
+#' * Will not retrieve a new OAuth token if it expires after the promised request
+#'   is created but before it is actually requested.
+#' * Does not perform throttling with [req_throttle()].
+#' * Does not attempt retries as described by [req_retry()].
+#' * Only consults the cache set by [req_cache()] when the request is promised.
+#'
+#' @inheritParams req_perform
+#' @inheritParams req_perform_parallel
+#'
+#' @return a [promises::promise()] object which resolves to a [response] if
+#' successful or rejects on the same errors thrown by [req_perform()].
+#' @export
+#'
+#' @examplesIf requireNamespace("promises", quietly = TRUE)
+#' library(promises)
+#' request_base <- request(example_url()) |> req_url_path_append("delay")
+#'
+#' p <- request_base |> req_url_path_append(2) |> req_perform_promise()
+#'
+#' # A promise object, not particularly useful on its own
+#' p
+#'
+#' # Use promise chaining functions to access results
+#' p %...>%
+#'   resp_body_json() %...>%
+#'   print()
+#'
+#'
+#' # Can run two requests at the same time
+#' p1 <- request_base |> req_url_path_append(4) |> req_perform_promise()
+#' p2 <- request_base |> req_url_path_append(1) |> req_perform_promise()
+#'
+#' p1 %...>%
+#'   resp_url_path %...>%
+#'   paste0(., " finished") %...>%
+#'   print()
+#'
+#' p2 %...>%
+#'   resp_url_path %...>%
+#'   paste0(., " finished") %...>%
+#'   print()
+#'
+#' # See the [promises package documentation](https://rstudio.github.io/promises/)
+#' # for more information on working with promises
+req_perform_promise <- function(req,
+                                path = NULL,
+                                pool = NULL) {
+  check_installed(c("promises", "later"))
+  check_string(path, allow_null = TRUE)
+
+  promises::promise(
+    function(resolve, reject) {
+      perf <- PerformancePromise$new(
+        req = req,
+        resolve = resolve,
+        reject = reject,
+        path = path,
+        error_call = environment()
+      )
+
+      perf$submit(pool)
+    }
+  )
+}
+
+PerformancePromise <- R6Class("PerformancePromise", inherit = Performance,
+  public = list(
+    resolve = NULL,
+    reject = NULL,
+
+    initialize = function(req, resolve, reject, path = NULL, error_call = NULL) {
+      progress <- create_progress_bar(config = FALSE)
+
+      super$initialize(req = req, path = path, progress = progress, error_call = error_call)
+      self$resolve <- resolve
+      self$reject <- reject
+    },
+
+    submit = function(pool = NULL) {
+      if (!is.null(self$resp)) {
+        # cached
+        self$resolve(self$resp)
+        return()
+      }
+      super$submit()
+      ensure_pool_poller(pool, self$reject)
+    },
+
+    succeed = function(res) {
+      tryCatch({
+        super$succeed(res)
+        self$resolve(self$resp)
+      },
+        httr2_fail = function(cnd) self$reject(cnd$error),
+        error = function(cnd) self$reject(cnd)
+      )
```

Comment:
Would you mind using this indenting style? I find it a bit easier to read.

```suggestion
      tryCatch(
        {
          super$succeed(res)
          self$resolve(self$resp)
        },
        httr2_fail = function(cnd) self$reject(cnd$error),
        error = function(cnd) self$reject(cnd)
      )
```